cosmos/config.py

```diff
@@ -257,6 +259,36 @@ def validate_profiles_yml(self) -> None:
         if self.profiles_yml_filepath and not Path(self.profiles_yml_filepath).exists():
             raise CosmosValueError(f"The file {self.profiles_yml_filepath} does not exist.")
 
+    def _handle_profile_caching(self, desired_profile_path: Path | None = None, use_mock_values: bool = False) -> Path:
+        """
+        Handle the profile caching mechanism.
+
+        Check if profile object version is store in metadata.json and return the profile path
+        If Version not found in metadata.json then store version in metadata.json and create
+        corresponding profile at given path.
+        """
+        # TODO: Fix type
+        profile_version = self.profile_mapping.version(use_mock_values)  # type: ignore[union-attr]
+        cached_profile_path = get_cache_profile_version_path(profile_version)
+        if cached_profile_path:
+            logger.info("Profile found in cache using profile: %s.", cached_profile_path)
+            return cached_profile_path
+        else:
+            if desired_profile_path:
+                profile_path = desired_profile_path
+            else:
+                profile_path = Path(dbt_profile_cache_path) / profile_version / DEFAULT_PROFILES_FILE_NAME
+                cache_profile_version_path(profile_version, profile_path.__str__())
+                # TODO: Fix type
+                profile_contents = self.profile_mapping.get_profile_file_contents(  # type: ignore[union-attr]
+                    profile_name=self.profile_name, target_name=self.target_name, use_mock_values=use_mock_values
+                )
+
+                Path(Path(dbt_profile_cache_path) / profile_version).mkdir(parents=True, exist_ok=True)
+                logger.info("Profile not found in cache storing and using profile: %s.", profile_path)
+                profile_path.write_text(profile_contents)
```

Comment:
With the current approach, we'd be storing all the caches for all the versions of profiles - regardless if they exist or not. Ideally, we would only keep the relevant cache to the latest definition of a specific profile, overriding it if the parameters changed.

Reply:
If the cache exists, i.e hash for the requested profile object exists in the cache path, then we are not rewriting it. 

> Ideally, we would only keep the relevant cache to the latest definition of a specific profile, overriding it if the parameters changed.

Could you please elaborate a bit more? Are you suggesting to keep only one profile cached per profile type? 

 was thinking that it makes sense to store in cache when object parameters change. For example, caching two profiles based on the object below, because the conn_id parameter has changed.

```
profile_mapping=PostgresUserPasswordProfileMapping(
        conn_id="airflow_db",
        profile_args={"schema": "public"},
        disable_event_tracking=True,
    ),
```

```
profile_mapping=PostgresUserPasswordProfileMapping(
        conn_id="airflow_db_1",
        profile_args={"schema": "public"},
        disable_event_tracking=True,
    ),
```